rust/worker/src/execution/operators/fetch_segment.rs

```diff
@@ -0,0 +1,207 @@
+    NoCollectionDimension,
+    #[error("No segment found")]
+    NoSegment,
+    #[error("Unable to create record segment reader: {0}")]
+    RecordSegmentReaderCreation(#[from] RecordSegmentReaderCreationError),
+    // The frontend relies on ths content of the error message here to detect version mismatch
+    // TODO: Refactor frontend to properly detect version mismatch
+    #[error("Collection version mismatch")]
+    VersionMismatch,
+}
+
+impl ChromaError for FetchSegmentError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            FetchSegmentError::GetCollection(e) => e.code(),
+            FetchSegmentError::GetSegment(e) => e.code(),
+            FetchSegmentError::HNSWConfigError(e) => e.code(),
+            FetchSegmentError::HNSWSegmentReaderCreation(e) => e.code(),
+            FetchSegmentError::MetadataSegmentReaderCreation(e) => e.code(),
+            FetchSegmentError::NoCollection => ErrorCodes::NotFound,
+            FetchSegmentError::NoCollectionDimension => ErrorCodes::InvalidArgument,
+            FetchSegmentError::NoSegment => ErrorCodes::NotFound,
+            FetchSegmentError::RecordSegmentReaderCreation(e) => e.code(),
+            FetchSegmentError::VersionMismatch => ErrorCodes::VersionMismatch,
+        }
+    }
+}
+
+impl FetchSegmentOutput {
+    pub(super) async fn knn_segment_reader(
+        &self,
+    ) -> Result<Option<DistributedHNSWSegmentReader>, FetchSegmentError> {
+        match DistributedHNSWSegmentReader::from_segment(
+            &self.knn,
+            self.collection
+                .dimension
+                .ok_or(FetchSegmentError::NoCollectionDimension)? as usize,
+            self.hnsw.clone(),
+        )
+        .await
+        {
+            Ok(reader) => Ok(Some(*reader)),
+            Err(e) => match *e {
+                DistributedHNSWSegmentFromSegmentError::Uninitialized => Ok(None),
+                e => Err(e.into()),
+            },
+        }
+    }
+
+    pub(super) fn knn_config(&self) -> Result<IndexConfig, FetchSegmentError> {
+        Ok(IndexConfig::from_segment(
+            &self.knn,
+            self.collection
+                .dimension
+                .ok_or(FetchSegmentError::NoCollectionDimension)?,
+        )
+        .map_err(|e| *e)?)
+    }
+
+    pub(super) async fn metadata_segment_reader(
+        &self,
+    ) -> Result<MetadataSegmentReader, FetchSegmentError> {
+        Ok(MetadataSegmentReader::from_segment(&self.metadata, &self.blockfile).await?)
+    }
+
+    pub(super) async fn record_segment_reader(
+        &self,
+    ) -> Result<Option<RecordSegmentReader>, FetchSegmentError> {
+        match RecordSegmentReader::from_segment(&self.record, &self.blockfile).await {
+            Ok(reader) => Ok(Some(reader)),
+            Err(e) => match *e {
+                RecordSegmentReaderCreationError::UninitializedSegment => Ok(None),
+                e => Err(e.into()),
+            },
+        }
+    }
+}
+
```

Comment:
we can remove alot of - https://github.com/chroma-core/chroma/blob/main/rust/worker/src/execution/orchestration/common.rs when this is finally all done